### PR TITLE
Unmarshal bulk result into codegen structs

### PR DESCRIPTION
**Pull request by rooftoofwoof**
This PR adds 

- [x] a `OnlineQueryBulkResult` method `UnmarshalInto` that takes in a pointer to a slice and deserializes the bulk result into a list of struct instances of the same element type as the input slice. This does NOT yet support unmarshalling has-manys. 
- [x] a `UnmarshalTableInto` function that can unmarshal a has-many table into its feature structs. This is introduced as a stopgap solution to `OnlineQueryBulkResult.UnmarshalInto` not currently handling deserialization of has-many features

### CODE REVIEW

File: internal/feather.go [hunk 1]
```diff
@@ -275,7 +327,6 @@ func recordToBytes(record arrow.Record) ([]byte, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to close Arrow Table writer: %w", err)
 	}
-	record.Release()
```
Comment (rooftoofwoof) [inline internal/feather.go:278]: caller's responsibility

File: unmarshal.go [hunk 2]
```diff
@@ -93,24 +95,105 @@ func (f fqnToFields) setFeature(fqn string, value any) error {
...
+//				Features.User.Relatives,
+//			).WithInput(Features.User.Id, []int{1, 2}), nil)
+//
+//			relatives := make([]Relative, 0)
+//			result.UnmarshalInto(&relatives)
+//
+//	     feature, _ := chalk.UnwrapFeature(Features.User.Relatives)
+//			fmt.Println("Number of relatives for all users: ", len(result.GroupsTable[feature.Fqn]))
+//
+//		}
```
Comment (rooftoofwoof) [inline unmarshal.go:178]: indent

File: internal/feather.go [hunk 3]
```diff
@@ -65,7 +109,8 @@ func inputsToArrowBytes(inputs map[string]any) ([]byte, error) {
 		if !ok {
 			return nil, fmt.Errorf("failed to find input values for feature '%s'", field.Name)
 		}
-		reflectKind := reflect.ValueOf(values).Type().Elem().Kind()
+		elem := reflect.ValueOf(values).Type().Elem()
```
Comment (Nathan-Fenner) [inline internal/feather.go:112]: nit: you can abbreviate `reflect.ValueOf(x).Type()` as `reflect.TypeOf(x)`. But also, it's probably best to confirm that `values` is really a slice type or you'll get a very confusing error later. Something like

```go
valuesType := reflect.TypeOf(values)
if valuesType.Kind() != reflect.Slice {
  return fmt.Errorf("the value provided for feature '%s' is not of slice type; it has type %v", field.Name, valuesType)
}
elemType := valuesType.Elem()
```

File: internal/feather.go [hunk 4]
```diff
@@ -105,15 +150,22 @@ func inputsToArrowBytes(inputs map[string]any) ([]byte, error) {
 			recordBuilder.Field(idx).(*array.LargeStringBuilder).AppendValues(values.([]string), nil)
 		case reflect.Bool:
 			recordBuilder.Field(idx).(*array.BooleanBuilder).AppendValues(values.([]bool), nil)
-		default:
-			if reflectKind.String() == "time.Time" {
-				// TODO: Support this
-				//recordBuilder.Field(idx).(*array.TimestampBuilder).AppendValues(values.([]time.Time), nil)
+		case reflect.Struct:
+			if elem.String() == "time.Time" {
```
Comment (Nathan-Fenner) [inline internal/feather.go:154]: It is more robust to instead check

```go
if elem == reflect.TypeFor[time.Time]() {
  ...
}
```

File: internal/feather.go [hunk 5]
```diff
@@ -105,15 +150,22 @@ func inputsToArrowBytes(inputs map[string]any) ([]byte, error) {
...
+				timeSlice := values.([]time.Time)
+				timestampSlice := make([]arrow.Timestamp, 0)
```
Comment (Nathan-Fenner) [inline internal/feather.go:156]: Should set the capacity here, with

```go
timestampSlice := make([]arrow.Timestamp, 0, len(timeSlice))
```

File: internal/unmarshal.go [hunk 6]
```diff
@@ -122,18 +134,127 @@ func getPointerToCopied(elemType reflect.Type, value any) reflect.Value {
...
+					for ptr := arr.Offsets()[i]; ptr < arr.Offsets()[i+1]; ptr++ {
+						anyVal, valueErr := GetValueFromArrowArray(arr.ListValues(), int(ptr))
+						if valueErr != nil {
+							return nil, fmt.Errorf("error getting value for LargeList column: %w", valueErr)
+						}
+						newSlice = append(newSlice, anyVal)
+					}
+					m[name] = newSlice
+				case *array.Struct:
+					// TODO: Deserialize structs to support dataclasses
```
Comment (Nathan-Fenner) [inline internal/unmarshal.go:206]: Should this at least return an error right now to indicate that it's unsupported?

File: unmarshal.go [hunk 7]
```diff
@@ -93,24 +94,105 @@ func (f fqnToFields) setFeature(fqn string, value any) error {
...
+	slice := reflect.Indirect(slicePtr)
+	if slice.Kind() != reflect.Slice && slice.Kind() != reflect.Array {
+		return fmt.Errorf(
+			"result holder should be a pointer to a slice of structs, "+
+				"got '%s' instead",
+			slice.Kind(),
+		)
+	}
+
+	sliceType := slice.Type().Elem()
```
Comment (Nathan-Fenner) [inline unmarshal.go:136]: nit: this is a confusing name, maybe `sliceElementType` instead?

File: unmarshal.go [hunk 8]
```diff
@@ -93,24 +94,105 @@ func (f fqnToFields) setFeature(fqn string, value any) error {
...
+	if slicePtr.Kind() != reflect.Ptr {
+		return fmt.Errorf(
+			"result holder should be a pointer to a slice of structs, "+
+				"got '%s' instead",
+			slicePtr.Kind(),
+		)
+	}
+
+	slice := reflect.Indirect(slicePtr)
+	if slice.Kind() != reflect.Slice && slice.Kind() != reflect.Array {
```
Comment (Nathan-Fenner) [inline unmarshal.go:128]: The code below `append`s into the slice, so the `reflect.Array` case won't actually work, and shouldn't be permitted here.

**Comment by Nathan-Fenner** [comment]
Generally looks good to me, but a few small things to fix before merging.

**Review by Nathan-Fenner — APPROVED** [review]

**Pull request merged**